### PR TITLE
Use some of the newly available petsc4py function / simplify

### PR DESCRIPTION
**Pull request by rbeucher**
Here is a first pass at simplifying some of the code.
The tests are passing with Petsc 3.17.2
we don't have great test coverage though so I can't guarantee that it's 100% safe.

I have more coming.

R

### CODE REVIEW

**Review by lmoresi — COMMENTED** [review]
Looks promising. Is the swarm.pyx now completely deprecated ?

File: underworld3/swarm.py [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+        kd.build_index()
+
+        with self.swarm.access():
+            n,d,b = kd.find_closest_point(self.swarm.data)
+   
+            node_values  = np.zeros((self._meshVar.coords.shape[0],self.num_components))
+            w = np.zeros(self._meshVar.coords.shape[0]) 
+
+            if not self._nn_proxy:
+                for i in range(self.data.shape[0]):
```
Comment (julesghub) [inline underworld3/swarm.py]: Can this be done without the `for` loop, i.e. pure numpy?

**Review by julesghub — APPROVED** [review]
Great work @rbeucher - I think this is all great stuff!
My only concerns are the `for` loops for the swarms. Happy to discuss further on that.

File: underworld3/swarm.py (hunk 1 above)
Comment (lmoresi) [reply underworld3/swarm.py]: No, this is the reduction operation (+=) and so breaks if not done carefully / sequentially
One way to do this is to accumulate all the neighbours into a 2d array and use numpy to sum them up. That does require knowing the maximum size of the array

**Comment by rbeucher** [comment]
I think we need to extend or test coverage before refactoring heavily.

File: underworld3/maths.pyx [hunk 2]
```diff
@@ -180,11 +173,7 @@ class CellWiseIntegral:
         cdef DS ds = dmc.getDS()
         CHKERRQ( PetscDSSetObjective(ds.ds, 0, ext.fns_residual[0]) )
         
-        cdef Vec rvec = dmc.createGlobalVec()
-        CHKERRQ( DMPlexComputeCellwiseIntegralFEM(dmc.dm, cgvec.vec, rvec.vec, NULL) )
+        results = dmc.DMPlexComputeCellwiseIntegralFEM(cgvec)
```
Comment (rbeucher) [inline underworld3/maths.pyx]: I realised this is not in petsc4py yet... I'll have to merge it first

**Comment by lmoresi** [comment]
I'd be happy to approve but can we move swarm.pyx out of the way altogether ?

**Comment by rbeucher** [comment]
Don't merge. There is more coming. It's just so you are aware.

**Comment by lmoresi** [comment]
All fine - I won't merge. I just updated to 3.17.2 which is the latest available via home brew so I would suggest we only merge functionality that is available in that version. 

I must admit, this is the best change - so much simplification and a handoff of lots of our responsibility to the community!

L

**Comment by rbeucher** [comment]
OK so this is PETSc 3.17.2 compatible. I am merging into dev.

**Pull request merged**